packages/insomnia/src/routes/ai.generate-commit-messages.tsx

```
@@ -0,0 +1,72 @@
+import { href } from 'react-router';
+
+import { showToast } from '~/ui/components/toast-notification';
+import { createFetcherSubmitHook } from '~/utils/router';
+
+import type { Route } from './+types/ai.generate-commit-messages';
+
+export async function clientAction(args: Route.ClientActionArgs) {
+  const { projectId } = (await args.request.json()) as { projectId: string };
+
+  try {
+    const { changes } = await window.main.git.gitChangesLoader({ projectId });
+    if (changes.staged.length > 0) {
+      return {
+        error: 'You have staged changes. Please commit or unstage them and try again.',
+      };
+    }
+    const diff = await window.main.git.diff();
+
+    console.log('Diff for AI:\n', diff);
+
+    const { log } = await window.main.git.gitLogLoader({ projectId });
+
+    const { error, commits } = await window.main.generateCommitsFromDiff({
+      diff,
+      recent_commits: log
+        .slice(0, 5)
+        .map(({ commit }) => commit.message)
+        .join('\n'),
+    });
+
+    if (error || !commits) {
+      showToast({
+        title: 'Failed to generate commit messages',
+        icon: 'star',
+        status: 'error',
+        description: 'The AI service returned invalid data. Please try again.',
+      });
+      return {
+        error: 'The AI service returned invalid data. Please try again.',
+      };
+    }
+
+    return {
+      commits: commits.map(commit => ({
+        id: crypto.randomUUID(),
+        ...commit,
+      })),
+    };
+  } catch (err) {
+    showToast({
+      title: 'Failed to generate commit messages',
+      icon: 'star',
+      status: 'error',
+      description: 'There was an error communicating with the AI service. Please try again.',
```

Should we surface the actual error? If it's an issue with something like the API key it would be preferable to surface that so it can be resolved before trying again

---

I appended the error to the message.
Since these errors can be very low level we might want to keep the user facing message simple and add the whole trace and complex error to the console instead.
Also we have been discussing on surfacing all app errors in a more user friendly error panel alongside some error codes/remediation strategies as future work

---

>Since these errors can be very low level we might want to keep the user facing message simple and add the whole trace and complex error to the console instead.

Yeah, I'm good with that too. The main problem I want to solve is giving the user enough information to be transparent about what is failing, in case it's something on their side.